### ISSUE TITLE
How to nest modules that update their own parameters?

### ISSUE THREAD

**Issue by danijar** [deepmind/dm-haiku#427]
I have a quite complex TF2 code base that I'm trying to convert to JAX / Haiku. There are many nested modules, some of which contain custom logic for updating their own parameters. Some of the modules are also used for computing the loss of other modules.

Is there some way to put the logic for parameter updates into the corresponding module, so that train functions can be nested, rather than orchestrating all updates from the outside (which becomes too complex)?

Here is an example that illustrates this idea of nesting train functions. However, it doesn't run because Haiku modules can only be instantiated inside of a `hk.transform`:

```python3
import haiku as hk
import jax
import jax.numpy as jnp
import optax
from tensorflow_probability.substrates import jax as tfp

tfd = tfp.distributions
sg = jax.lax.stop_gradient


class Outer(hk.Module):

  def __init__(self, data_size, code_size):
    self.inner = Inner(data_size, code_size)
    self.tloss = hk.transform(self.loss)

  def build(self, x1, x2):
    params = {}
    params['inner'] = self.inner.init(x1)
    params['pred'] = self.tloss.init(x2)
    return params

  def train(self, params, x1, x2):
    params['inner'] = self.inner.train(params['inner'], x1)  # <---
    grad = jax.grad(self.tloss.apply)(params['pred'], x2)
    params['pred'] = jax.tree_map(
        lambda w, g: w - 0.1 * g, params['pred'], grad)
    return params

  def loss(self, x):
    target = sg(self.inner.encode(x))
    return ((self.pred(x) - target) ** 2).mean()

  def pred(self, x):
    x = jax.nn.relu(hk.Linear(128)(x))
    x = hk.Linear(self.inner.code_size)(x)
    return x


class Inner(hk.Module):

  def __init__(self, data_size, code_size):
    self.data_size = data_size
    self.code_size = code_size
    self.tloss = hk.transform(self.loss)
    mean, std = jnp.zeros(code_size), jnp.ones(code_size)
    self.prior = tfd.Independent(tfd.Normal(mean, std), 1)

  def build(self, x):
    return self.tloss.init(x)

  def train(self, params, x):
    grad = jax.grad(self.tloss.apply)(params, x)
    return optax.apply_update(params, grad)
    return jax.tree_map(lambda w, g: w - 0.1 * g, params, grad)

  def loss(self, x):
    post = self.encode(x)
    recon = self.decode(post.sample())
    mse = ((recon - x) ** 2).sum(-1).mean()
    kl = post.kl_divergence(self.prior).mean()
    return mse + 0.1 * kl

  def encode(self, x):
    x = jax.nn.relu(hk.Linear(128)(x))
    x = jax.nn.relu(hk.Linear(128)(x))
    mean = hk.Linear(self.code_size)(x)
    std = jax.nn.softplus(hk.Linear(self.code_size)(x))
    return tfd.Independent(tfd.Normal(mean, std), 1)

  def decode(self, x):
    x = jax.nn.relu(hk.Linear(128)(x))
    x = jax.nn.relu(hk.Linear(128)(x))
    return hk.Linear(self.data_size)(x)


def main():
  x1 = jnp.zeros((64, 128), jnp.float32)
  x2 = jnp.zeros((64, 32), jnp.float32)
  module = Outer(128, 8)
  params = module.build(x1, x2)
  tpred = hk.transform(module.pred)
  params = jax.jit(module.train)(params, x1, x2)
  y = jax.jit(tpred.apply)(params, x2)
  print(y)


if __name__ == '__main__':
  main()
```
